src/primitives/transaction.h

```diff
@@ -351,11 +357,39 @@ class CTransaction
 
     template <typename Stream, typename Operation>
     inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
-        READWRITE(*const_cast<int32_t*>(&this->nVersion));
+        if (ser_action.ForRead()) {
+            // When deserializing, unpack the 4 byte header to extract fOverwintered and nVersion.
+            uint32_t header;
+            READWRITE(header);
+            *const_cast<bool*>(&fOverwintered) = header >> 31;
+            *const_cast<int32_t*>(&this->nVersion) = header & 0x7FFFFFFF;
+        } else {
+            if (fOverwintered) {
+                // When serializing Overwintered tx, assemble the 4 byte header from fOverwintered and nVersion
+                uint32_t header = (1 << 31) | this->nVersion;
+                READWRITE(header);
+            } else {
+                // When serializing v1 and v2, the 4 byte header is nVersion
+                READWRITE(*const_cast<int32_t*>(&this->nVersion));
+            }
+        }
         nVersion = this->nVersion;
+        if (fOverwintered) {
+            READWRITE(*const_cast<uint32_t*>(&this->nFormatBranchId));
+        }
+        // FIXME: Do we want to do this?  Will it lead to unexpected code paths?
+        if (fOverwintered && nVersion != 3) {
+            throw std::ios_base::failure("Only v3 format Overwintered txs are currently supported");
+        }
+        bool isOverwinterV3 = fOverwintered &&
+                              nVersion == 3 &&
+                              nFormatBranchId == NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId;
```

Comment:
Move this `nFormatBranchId` into its own check (probably using the same `tx.nVersion -> tx.nFormatBranchId` mapping as in the consensus rules) immediately after its `READWRITE` call, and throw an exception if the mapping is invalid.

Reply:
Going to leave as is for now, can't share code as CMutableTransaction is a struct and not a subclass of CTransaction.